### ISSUE TITLE
Implement generics

### ISSUE THREAD

**Issue by cosmos72** [gopherdata/gophernotes#229]
Now that Go generics have been approved, we should support them in gophernotes too

More specifically, the underlying interpreter gomacro has an open issue https://github.com/cosmos72/gomacro/issues/24 to track their implementation.

Help is welcome :)

**Comment by cosmos72** [comment]
@dwhitena @SpencerPark  @sbinet any idea or suggestion on the best approach to move forward ?

**Comment by sbinet** [comment]
I don't remember whether gomacro is using reflect under the hood to manipulate types, but if it does, one could perhaps weigh in and propose to have a reflect API that can also manipulate parametrized types.

**Comment by cosmos72** [comment]
Yes, gomacro heavily uses `reflect` under the hood to manipulate types and values.

This has several limitations, including that `reflect` package cannot create named types or interface types.

Generics are not supported by `reflect`, and if I remember correctly there's no plan to support them. Only **instantiated** generic types will be supported.

And I am quite sure there's no plan to support instantiating generics at runtime - which for the foreseeable future will require invoking the Go toolchain.

**Comment by cosmos72** [comment]
A third solution came to my mind: when importing a package, instantiate its generics on a custom type that satisfies the constraint(s).

First example: for the `Stack[T]` above, instantiate `Stack[interface{}]` when importing, compiling and loading `example.com/go/stack` with Go toolchain.

Second example: for an hypothetical generic type
```Go
type Set[T Ordered] { /* ... */ }
```
where the constraint `Ordered` is
```
type Ordered[T] interface {
    Compare(T other) int
}
```
we can instantiate, compile and load with Go toolchain the type `Set[OrderedX]` where `OrderedX` is
```Go
type OrderedX {
  X_ interface{}
  Compare_ func(other OrderedX) int
}
// implement the interface Ordered[OrderedX]
func (obj OrderedX) Compare(other OrderedX) int {
    return obj.Compare_(other)
}
```
